satpy/readers/generic_image.py

```diff
@@ -117,4 +136,24 @@ def mask_image_data(data):
                                 for i in range(data.shape[0])])
         data.data = masked_data
         data = data.sel(bands=BANDS[data.bands.size - 1])
+    elif hasattr(data, 'nodatavals') and data.nodatavals:
+        try:
+            nodata_handling = info['nodata_handling'] if info else NODATA_HANDLING_FILLVALUE
+        except KeyError:
+            nodata_handling = NODATA_HANDLING_FILLVALUE
+        if nodata_handling == NODATA_HANDLING_NANMASK:
+            # data converted to float and masked with np.nan
+            data = data.astype(np.float32)
+            masked_data = da.stack([da.where(data.data[i, :, :] == nodataval, np.nan, data.data[i, :, :])
+                                    for i, nodataval in enumerate(data.nodatavals)])
+            data.data = masked_data
+            data.attrs['_FillValue'] = np.nan
+        elif nodata_handling == NODATA_HANDLING_FILLVALUE:
+            # keep data as it is but set _FillValue attribute to provided
+            # nodatavalue (first one as it has to be the same for all bands at least
+            # in GeoTiff, see GDAL gtiff driver documentation)
+            fill_value = data.nodatavals[0]
+            if np.issubdtype(data.dtype, np.integer):
+                fill_value = int(fill_value)
```

Comment:
Is the nodatavals ever not an integer? What case is this if statement handling? Mostly just curious if it is needed and could be removed.

Reply:
Yes, seems that rasterio returns this as float - at least in my setup (nodata value 0 was returns as `[0.0,]`).

Reply:
Ok. The only other issue I could see this having is if `int(fill_value)` is producing a 64-bit integer and is used in later numpy/xarray operations if it ends up upcasting the array data to 64-bit. I don't think this is true, but want to mention it here in case you run into it in your future changes.